manual/wavesfc.py.1.ronn
# wavesfc.py(1) -- from arrival times and a trilaterated model, produce an irregular surface mesh


## SYNOPSIS

`wavesfc.py` <CONFIGURATION>


## DESCRIPTION

`wavesfc.py` uses the center, sound speed and radius recovered from a model
trilaterated by `trilateration.py`, along with the backscatter arrival times
estimated by `atimes.py`, to develop an irregular surface model represented as
a Delaunay triangulation. Along the straight from each element to the
center of a trilaterated model, the backscatter arrival time and exterior
sound speed are used to determine the round-trip propagation distance that
element. Half of this round-trip propagation distance is interpreted as the
position of a mesh node along the element-to-center line.


## CONFIGURATION

The YAML configuration file (or the rendered Mako template that produces valid
YAML) should define a dictionary with four sections. Each section is, in turn,
a dictionary that defines configuration options. Some sections are common to
multiple programs, allowing these programs to share a single configuration
file without duplicating common options.

The `general` section can include one relevant key:

- `include`:
  A single string or a list of strings providing paths to other YAML
  configuration files that can be used to import options. Once set, an option
  is not replaced by an include. Thus, an option set in a top-level file will
  not be overridden by an option in an included file. When multiple includes
  are specified, options are set from each include file in the order the files
  are listed.

The `measurement` section include two relevant keys:

- `radius`:
  The initial estimate for the radius of the target (or several identical targets).

- `c`:
  The initial estimate for the sound speed of the background medium surrounding
  the target (or several identical targets).

The `wavesfc` section includes several options:

- `timefile`:
  A single path or glob, or a list of paths or globs, to keymat(7) files that
  map transmit-receive pairs to arrival times. Any transmit-receive pair for
  which the transmit and receive indices are not equal are discarded. If each
  transmit-receive pair maps to a list of more than one arrival time, one
  target positions per arrival time will be determined.

- `elements`:
  A keymat(7) file that maps individual element indices (as single integers) to
  a location as an (x, y, z) array of floats.

- `reflectors`:
  A text file representing an N-by-M matrix, where N is the number of targets
  (arrival times) for each measurement and M is 3, 4 or 5. If M is 3, the row
  specificies the (x, y, z) coordinates of the center of the target. The
  exterior radius and sound speed for that target will use the values specified
  in `radius` and `c` values in the `measurement` section, respectively. If M
  is 4, the fourth column specifies a sound speed that overrides the
  `measurement` value and may be unique to each target location. If M is 5, the
  fifth column specifies a radius that overrides the `measurement` value and
  may be unique to each target location.

- `output`:
  The name of a file that will be written to store the Delaunay triangulation
  of the surface, as a Numpy *npz* file with constituent files `nodes`,
  `triangles`, and `elements`. The `nodes` and `triangles` arguments are
  standard mesh parameters. Because the mesh may not contain nodes derived from
  some elements that contain missing or bad arrival times, each row of (x, y,
  z) coordinates in the `nodes` array has a corresponding row in the `elements`
  array that indicates the index of the element used to determine that node.

  If `reflectors` specifies more than a single target, one mesh file will be
  written for each target. A string 'Target`<INDEX>`' will be prepended to any
  extension in the output file, where `<INDEX>` indicates the corresponding row
  in `reflectors` (starting from zero). The width of the index field will be
  dynamically chosen based on the total number of targets and zero-padded as
  necessary.

- `usemiss` (optional, default _False_):
  If _True_, any nodes that cannot be placed because backscatter times for the
  corresponding element are either missing or discarded as outliers will be
  placed by a node on the hemispheric model implied by the trilteration. This
  avoids very large patches in the mesh caused by missing nodes, at the expense
  of forcing a hemispheric representation in the neighborhood of those nodes.

- `olgroup` (optional, default _64_):
  For arrival times corresponding to element indices in contiguous blocks of
  size `olgroup`, determine the interquartile range and discard outlier arrival
  times (see `olrange`).

- `olrange` (optional, default _None_):
  If specified, an arrival time in each contiguous group of `olgroup` elements
  is considered an outlier and excluded if it falls more than `olrange` IQR
  above the third quartile or below the first quartile. If `olrange` is
  omitted, no outliers are excluded.


## COPYRIGHT

Copyright (c) 2014--2018 Andrew J. Hesford. All rights reserved.


## SEE ALSO

atimes.py(1) - For a descrption of arrival-time estimation  
trilateration.py(1) - For a description of the trilateration process  
keymat(7) - For a description of the keymat file format

[Mako Templates for Python](http://www.makotemplates.org)


shell/wavesfc.py
#!/usr/bin/env python

# Copyright (c) 2015 Andrew J. Hesford. All rights reserved.
# Restrictions are listed in the LICENSE file distributed with this package.

import os, sys, numpy as np
from numpy.linalg import norm

from scipy.spatial import Delaunay

from collections import defaultdict

from pycwp import stats, cutil

from habis.habiconf import HabisConfigError, HabisConfigParser, matchfiles
from habis.formats import loadmatlist, savez_keymat, savetxt_keymat

def usage(progname):
	print('USAGE: %s <configuration>' % progname, file=sys.stderr)


def wavesfcEngine(config):
	'''
	Use positions of elements and targets specified in the provided config,
	combined with a specified sound speed (to override per-reflector sound
	speeds), to compute control points that define a reflector surface.
	'''
	msection = 'measurement'
	wsection = 'wavesfc'
	try:
		# Try to grab the input and output files
		eltfiles = matchfiles(config.getlist(wsection, 'elements'))
		rflfile = config.get(wsection, 'reflectors')
		outfile = config.get(wsection, 'output')
	except Exception as e:
		err = 'Configuration must specify elements, reflectors and output in [%s]' % wsection
		raise HabisConfigError.fromException(err, e)

	try:
		# Try to read the input time files
		timefiles = matchfiles(config.getlist(wsection, 'timefile'))
	except Exception as e:
		err = 'Configuration must specify timefile in [%s]' % wsection
		raise HabisConfigError.fromException(err, e)

	try:
		# Grab the outlier range and group size
		olrange = config.get(wsection, 'olrange', mapper=float, default=None)
		olgroup = config.get(wsection, 'olgroup', mapper=int, default=64)
	except Exception as e:
		err = 'Invalid specification of optionals outrange or outgroup in [%s]' % wsection
		raise HabisConfigError.fromException(err, e)

	try:
		# Grab the sound speed
		c = config.get(msection, 'c', mapper=float)
		rad = config.get(msection, 'radius', mapper=float)
	except Exception as e:
		err = 'Configuration must specify c and radius in [%s]' % msection
		raise HabisConfigError.fromException(err, e)

	try:
		# Use the reflector radius for missing arrival-time values
		usemiss = config.get(wsection, 'usemiss', mapper=bool, default=False)
	except Exception as e:
		err = 'Invalid specification of optional usemiss in [%s]' % wsection
		raise HabisConfigError.fromException(err, e)

	try:
		# Use the reflector radius if it is larger than time-imputed radius
		maxrad = config.get(wsection, 'maxrad', mapper=bool, default=False)
	except Exception as e:
		err = 'Invalid specification of optional maxrad in [%s]' % wsection
		raise HabisConfigError.fromException(err, e)

	# Read the element positions and backscatter arrival times
	elements = loadmatlist(eltfiles, nkeys=1)
	times = { k[0]: v
			for k, v in loadmatlist(timefiles, scalar=False, nkeys=2).items()
			if k[0] == k[1] }

	reflectors = np.loadtxt(rflfile, ndmin=2)
	nrefl, nrdim = reflectors.shape

	# Make sure the reflector specification includes speed and radius
	if nrdim == 3:
		reflectors = np.concatenate([reflectors, [[c, rad]] * nrefl], axis=1)
	elif nrdim == 4:
		reflectors = np.concatenate([reflectors, [[rad]] * nrefl], axis=1)
	elif nrdim != 5:
		raise ValueError('Reflector file must contain 3, 4 or 5 columns')

	# Split the file name and extension
	fbase, fext = os.path.splitext(outfile)

	# Add a target specifier to outputs for multiple targets
	nrefl = len(reflectors)
	if nrefl == 1: idfmt = ''
	else: idfmt = '.Target{{0:0{width}d}}'.format(width=cutil.numdigits(nrefl))

	for ridx, refl in enumerate(reflectors):
		pos = refl[:3]
		lc, rad = refl[3:]

		# Convert all times to distances
		pdists = { }
		for el, elpos in elements.items():
			# Ray from reflector center to element
			dl = elpos - pos
			# Find distance to element and normalize ray
			ll = norm(dl)
			dl /= ll

			try:
				# Convert arrival time to distance from center
				ds = ll - times[el][ridx] * lc / 2.0
			except (KeyError, IndexError, TypeError):
				# Use default radius or skip missing value
				if usemiss: ds = rad
				else: continue
			else:
				if maxrad: ds = max(rad, ds)

			# Valid points will be inside segment or opposite reflector
			if ds <= ll:
				# Record distance and ray
				pdists[el] = (ds, dl)

		if olrange is not None:
			# Group distances according to olgroup
			pgrps = defaultdict(dict)
			for el, (ds, _) in pdists.items():
				pgrps[int(el / olgroup)][el] = ds
			# Filter outliers from each group and flatten map
			pdists = { el: pdists[el] for pg in pgrps.values()
					for el in stats.mask_outliers(pg, olrange) }

		# Sort remaining values, separate indices and control points
		cpel = sorted(pdists.keys())
		cpts = np.array([ pos + pdists[el][0] * pdists[el][1] for el in cpel ])

		fname = fbase + idfmt.format(ridx) + fext

		# Project control points for tesselation
		zmin = np.min(cpts[:,-1])
		zmax = np.max(cpts[:,-1])

		if np.allclose(zmin, zmax):
			# For flat structures, just use x-y
			tris = Delaunay(cpts[:,:2])
		else:
			# Otherwise, project radially
			zdiff = zmax - zmin
			prctr = np.mean(cpts, axis=0)
			prctr[-1] = zmax + zdiff
			prref = np.array(prctr)
			prref[-1] = zmin

			l = cpts - prctr[np.newaxis,:]
			d = (prref - prctr)[-1] / l[:,-1]

			if np.any(np.isinf(d)) or np.any(np.isnan(d)):
				raise ValueError('Failure in projection for tesselation')

			ppts = prctr[np.newaxis,:] + d[:,np.newaxis] * l
			tris = Delaunay(ppts[:,:2])

		# Save the control points and triangles
		np.savez(fname, nodes=cpts, triangles=tris.simplices, elements=cpel)


if __name__ == '__main__':
	if len(sys.argv) != 2:
		usage(sys.argv[0])
		sys.exit(1)

	# Read the configuration file
	try:
		config = HabisConfigParser(sys.argv[1])
	except:
		print('ERROR: could not load configuration file %s' % sys.argv[1], file=sys.stderr)
		usage(sys.argv[0])
		sys.exit(1)

	# Call the calculation engine
	wavesfcEngine(config)
